cmake: Increase cmake policy version

ryanofsky opened the pull request:
Increase cmake policy version from 3.12 to 4.1 to stop using old and deprecated CMake policies in standalone builds.

Also stop overriding policy version if a cmake parent project has already set one, so parent projects are able to control which policies are enabled by default based on their own needs. Specifically, in the Bitcoin Core subtree, this change causes the libmultiprocess cmake policy version to increase from 3.12 to 3.22, which is the version Bitcoin Core uses.

This PR also adds a new `newdeps` CI job to test CMake 4.1 and the master branch of Cap'n Proto. This PR doesn't change the minimum version of cmake required to build the project, which is still 3.12.

---

This PR is an alternative to #163 which increases the policy version to 3.31 but doesn't include fixes for CI jobs, and doesn't allow the bitcoin core build to choose a lower policy version. This PR is also an alternative to #175 which sets the policy version to 3.22 like the bitcoin build, but also causes builds with earlier versions of cmake to fail.

ryanofsky committed:
ci: add newdeps job testing newest versions of cmake and capnproto

ryanofsky committed:
cmake: Increase cmake policy version

Increase cmake policy version from 3.12 to 4.1 in standalone builds to stop
using very old and deprecated CMake policies in standalone builds.

Also stop overriding policy version if a parent project has already set one, so
parent projects are able to control which policies are used by default based on
their own practices and needs.

In the Bitcoin Core subtree, this change causes the libmultiprocess policy
version to increase from 3.12 to 3.22, which is the version Bitcoin Core sets.

This commit only changes the policy version, it does not change the specified
minimum version of cmake required to build the project, which is 3.12.

ryanofsky commented:
<!-- begin push-2 -->
Updated 6e29c02ec5afdb86874e83aa46c93b0d88c27968 -> 233b394123fcb96c5164dc1f4b1876b8d0ac6514 ([`pr/policy.1`](https://github.com/ryanofsky/libmultiprocess/commits/pr/policy.1) -> [`pr/policy.2`](https://github.com/ryanofsky/libmultiprocess/commits/pr/policy.2), [compare](https://github.com/ryanofsky/libmultiprocess/compare/pr/policy.1..pr/policy.2))<!-- end --> with a few fixes for CI errors

hebasto commented on CMakeLists.txt:
```diff
@@ -2,7 +2,33 @@
+  # The purpose of the policy version is to opt out of policies introduced in
+  # newer versions of CMake until they have been tested. If this number is
+  # changed, the version in ci/configs/newdeps.bash should be changed to match.
+  cmake_minimum_required(VERSION 3.12...4.1 FATAL_ERROR)
```
`FATAL_ERROR` will probably be [ignored](https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html).

ryanofsky committed:
cmake: fix find_package(Threads REQUIRED) error with new cmake policies

With latest cmake policies, find_package(Threads REQUIRED) fails in freebsd and
openbsd builds. The specific failures in these jobs happens due to the CMP0155
policy turning CMAKE_CXX_SCAN_FOR_MODULES on, which is turned off in the next
commit. But there are other things that could cause the threads package not to
be found, and there are platforms where it may not be required, so it is better
to make it an optional instead of required dependency.

Technically this change is not needed to make all CI jobs pass as long as the
next commit is present. But since this find_package error obscures other errors
that would be clearer, and since there are other conditions that could trigger
it, it is worth fixing generally. For example, this same failure also seems to
happen in the llvm job as well when CMP0137 is disabled or
CMAKE_TRY_COMPILE_NO_PLATFORM_VARIABLES is set to true.

Error looks like:

 + cmake /home/runner/work/libmultiprocess/libmultiprocess -G Ninja
-- The CXX compiler identification is Clang 16.0.6
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - no
CMake Error at /usr/local/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:233 (message):
  Could NOT find Threads (missing: Threads_FOUND)
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:603 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/share/cmake/Modules/FindThreads.cmake:226 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:41 (find_package)

and inside the CMakeConfigureLog.yaml file there are "/bin/sh:
CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS-NOTFOUND: not found" errors.

ryanofsky committed:
ci: add cmake debug output and --parallel build option

ryanofsky committed:
cmake: fix CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS-NOTFOUND error with new cmake policies

With latest cmake policies, openbsd and freebsd CI jobs fail due to a lack of a
clang-scan-deps tool. The tool could potentially be installed on these
platforms but it is unclear how to do that and the project isn't using modules
anyway, so just disable them here. Errors look like:

+ cmake --build . --parallel -t all tests mpexamples -- -k 0
[1/114] Scanning /home/runner/work/libmultiprocess/libmultiprocess/src/mp/util.cpp for CXX dependencies
FAILED: CMakeFiles/mputil.dir/src/mp/util.cpp.o.ddi
"CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS-NOTFOUND" -format=p1689 -- /usr/bin/c++  -I/home/runner/work/libmultiprocess/libmultiprocess/include -I/home/runner/work/libmultiprocess/libmultiprocess/build-openbsd/include -isystem /usr/local/include -Werror -Wall -Wextra -Wpedantic -Wno-unused-parameter -std=gnu++20 -x c++ /home/runner/work/libmultiprocess/libmultiprocess/src/mp/util.cpp -c -o CMakeFiles/mputil.dir/src/mp/util.cpp.o -resource-dir "/usr/lib/clang/16" -MT CMakeFiles/mputil.dir/src/mp/util.cpp.o.ddi -MD -MF CMakeFiles/mputil.dir/src/mp/util.cpp.o.ddi.d > CMakeFiles/mputil.dir/src/mp/util.cpp.o.ddi.tmp && mv CMakeFiles/mputil.dir/src/mp/util.cpp.o.ddi.tmp CMakeFiles/mputil.dir/src/mp/util.cpp.o.ddi
/bin/sh: CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS-NOTFOUND: not found

hebasto commented on ci/scripts/ci.sh:
```diff
@@ -21,17 +21,43 @@ cmake --version
 if ver_ge "$cmake_ver" "3.15"; then
-  cmake --build . -t "${BUILD_TARGETS[@]}" -- "${BUILD_ARGS[@]+"${BUILD_ARGS[@]}"}"
+  cmake --build . --parallel -t "${BUILD_TARGETS[@]}" -- "${BUILD_ARGS[@]+"${BUILD_ARGS[@]}"}"
 else
   # Older versions of cmake can only build one target at a time with --target,
   # and do not support -t shortcut
   for t in "${BUILD_TARGETS[@]}"; do
-    cmake --build . --target "$t" -- "${BUILD_ARGS[@]+"${BUILD_ARGS[@]}"}"
+    cmake --build . --parallel --target "$t" -- "${BUILD_ARGS[@]+"${BUILD_ARGS[@]}"}"
```
6ba1050bef0b5ea5df35d3623c9e101cceb82677

If the [`--parallel` option](https://cmake.org/cmake/help/latest/manual/cmake.1.html#build-a-project) is used without an argument:
> the native build tool's default number is used.

This means that the build will be efficiently parallelized with the "Ninja" generator, but not with the "Unix Makefiles" generator. However, not all scripts in `ci/configs/` pass `-G Ninja`.

hebasto commented on ci/scripts/ci.sh:
```diff
@@ -21,17 +21,43 @@ cmake --version
+  # Could add --trace / --trace-expand here too but they are very verbose.
+  cmake_args+=(--debug-find --debug-output --debug-trycompile --log-level=DEBUG)
+  cmake "$src_dir" "${cmake_args[@]}" || : "cmake exited with $?"
+  cat CMakeFiles/CMakeConfigureLog.yaml || true
```
6ba1050bef0b5ea5df35d3623c9e101cceb82677

This log file name has been used only since CMake 3.26.

ryanofsky commented:
re: https://github.com/bitcoin-core/libmultiprocess/pull/209#issuecomment-3292167145

> I am not happy with this. The title of the topic says "Increase cmake policy version", which I would expect to be a one line change. Instead, it has +95/-17 changes.

Happy to change the title if you have a suggestion. This PR is only increasing the policy version and improving the CI, and I did not feel the CI changes were were mentioning since they are supporting changes, not the goal of the PR.

>     * the  length of the bash buildscript is doubled. Why isn't this written in a portable cmake script? Why is CMake preferred over handwritten Makefiles when the actual CI logic is written in bash?

I think the language question should be raised new issue if you want to discuss it. (But just to provide a little context here if it can avoid an argument about the language of a ~70 line script, I just personally find bash a lot nicer and more useful than cmake and am much familiar with it. More generally, bash is commonly used in CI, there many more developers who know bash than cmake.)

The script does increase from 37 to 67 lines here to do two things:

1. To support cloning and building unreleased versions of cap'n proto, so we can fix or report compatibility issues upstream cap'n proto earlier, and not be blindsided by new changes.
2. To provide better output when find_package fails, because by default the information cmake provides is not useful for debugging.

IMO both changes are worth making and do not have any real downsides.

> If you have to explain what you are doing, you are doing it wrong.

Interesting point. It might be helpful if you could explain practical downsides of the change. Unless you also believe that if you have to explain why a change is wrong, then it must be right?

> Putting Threads::Threads behind a non-namespaced abstraction mpdeps makes the build less robust.

Can you explain this a little more as well? Is there a practical downside here or a different approach you would recommend?

ryanofsky commented:
re: https://github.com/bitcoin-core/libmultiprocess/pull/209#issuecomment-3292177333

> I'm struggling to create a MRE for the issue with the `FindThreads` module on FreeBSD. The following works just fine:

The error happens when you don't have `clang-scan-deps` installed on the build machine, as seems to be the case currently with our freebsd and openbs vms. This issue is not specifically related to the `find_package(Threads REQUIRED)` call but the `find_package(Threads REQUIRED)` call makes it harder to understand what is wrong because it obscures the real error. Making the threads package optional produces a clearer error later, and I think it's just more correct anyway since modern platforms don't use separate threading libraries.

If it helps there is a failing CI run with more complete output:

https://github.com/ryanofsky/libmultiprocess/actions/runs/17622518177/job/50071319558

And the relevant cmake policy is https://cmake.org/cmake/help/latest/policy/CMP0155.html

ryanofsky commented:
> Make sure the `CMakeLists.txt` files are declarative and follow best practices instead of inventing new approaches and deviations.

To be sure, I agree with you it would be better if Bitcoin Core followed cmake's recommended best practices and used new policies instead of setting an really old policy version.

But bitcoin core isn't doing that, and other bitcoin core developers have given reasons why it shouldn't do that, and while I don't agree with these reasons, I think it is a reasonable judgement call, and they are just choosing a different tradeoffs than I would. This PR makes libmultiprocess compatible with either approach for choosing a policy version and stops using really old policies (older than the ones bitcoin core chooses) in any case.

ryanofsky commented:
> Here is a [build log](https://github.com/hebasto/github-actions/actions/runs/17734446515/job/50392701817) from GHA using FreeBSD:

Is that link right? It seems to be a 404 for me. I also don't think I understand what it indicates if find_package(Threads REQUIRED) works in another build. We already know it works in many builds. You can see some builds where it is failing
in https://github.com/bitcoin-core/libmultiprocess/pull/209#issuecomment-3271994909 and https://github.com/ryanofsky/libmultiprocess/actions/runs/17622518177/job/50071319558. Making the threads library optional improves error output, works well and doesn't have other downsides, and makes sense conceptually on modern platforms that do not use separate libraries for threading.

hebasto commented:
> > Here is a [build log](https://github.com/hebasto/github-actions/actions/runs/17734446515/job/50392701817) from GHA using FreeBSD:
> 
> Is that link right? It seems to be a 404 for me. I also don't think I understand what it indicates if find_package(Threads REQUIRED) works in another build.

Sorry. I've made that repo public. The link should work now.

> We already know it works in many builds. You can see some builds where it is failing in [#209 (comment)](https://github.com/bitcoin-core/libmultiprocess/pull/209#issuecomment-3271994909) and https://github.com/ryanofsky/libmultiprocess/actions/runs/17622518177/job/50071319558. Making the threads library optional improves error output, works well and doesn't have other downsides, and makes sense conceptually on modern platforms that do not use separate libraries for threading.

I want to report the issue upstream and I need an MRE for it.

hebasto commented:
@ryanofsky 
> Some things I might suggest:
> 
>    * Starting from my failing commit and deleting unnecessary steps to find a minimal failing case

Thank you for the suggestion.

The culprit turned out to be the "Ninja" generator, which was quite surprising to me.

ryanofsky reviewed (commented):
Thanks for the review! Note: first two commits are now part of #212

---

<!-- begin push-7 -->
Rebased 1c2958bfffd60c8cf9ec9715699c1496f6377922 -> f38a2f794717fb4e926d81c805bc9dde06ed6989 ([`pr/policy.6`](https://github.com/ryanofsky/libmultiprocess/commits/pr/policy.6) -> [`pr/policy.7`](https://github.com/ryanofsky/libmultiprocess/commits/pr/policy.7), [compare](https://github.com/ryanofsky/libmultiprocess/compare/pr/policy.6-rebase..pr/policy.7))<!-- end --> due to conflicts with #312, #325, #327, #330, and #334
<!-- begin push-8 -->
Rebased f38a2f794717fb4e926d81c805bc9dde06ed6989 -> 2c86c5aa0b179c5009fba2f86af1efb535882f83 ([`pr/policy.7`](https://github.com/ryanofsky/libmultiprocess/commits/pr/policy.7) -> [`pr/policy.8`](https://github.com/ryanofsky/libmultiprocess/commits/pr/policy.8), [compare](https://github.com/ryanofsky/libmultiprocess/compare/pr/policy.7-rebase..pr/policy.8))<!-- end --> dropping unneeded changes and fixing CMAKE_REQUIRED_LIBRARIES assignment to fix HAVE_PTHREAD_GETNAME_NP check to fix IWYU errors https://github.com/bitcoin-core/libmultiprocess/actions/runs/31822988343/job/94840422890?pr=209
